miden-lib/asm/sat/asset.masm

```diff
@@ -0,0 +1,84 @@
+use.miden::sat::internal::account->internal_account
+use.miden::sat::internal::asset
+use.miden::sat::account
+
+# CONSTANTS
+# =================================================================================================
+
+#! Builds a fungible asset for the specified fungible faucet and amount.
+#!
+#! Stack: [faucet_id, amount]
+#! Output: [ASSET]
+#!
+#! - faucet_id is the faucet to create the asset for.
+#! - amount is the amount of the asset to create.
+#! - ASSET is the built fungible asset.
+export.build_fungible_asset
+    # assert the faucet is a fungible faucet
+    dup exec.internal_account::is_fungible_faucet assert
+    # => [faucet_id, amount]
+
+    # assert the amount is valid
+    dup.1 exec.asset::get_fungible_asset_max_amount lte assert
+    # => [faucet_id, amount]
+
+    # create the asset
+    push.0 swap push.0 swap
+    # => [ASSET]
+end
+
+#! Creates a fungible asset for the faucet the transaction is being executed against.
+#!
+#! Stack: [amount]
+#! Output: [ASSET]
+#!
+#! - amount is the amount of the asset to create.
+#! - ASSET is the created fungible asset.
+export.create_fungible_asset
+    # fetch the id of the faucet the transaction is being executed against.
+    exec.account::get_id
+    # => [id, amount]
+
+    # build the fungible asset
+    exec.build_fungible_asset
+    # => [ASSET]
+end
+
+#! Builds a non fungible asset for the specified non-fungible faucet and amount.
+#!
+#! Stack: [faucet_id, DATA_HASH]
+#! Output: [ASSET]
+#!
+#! - faucet_id is the faucet to create the asset for.
+#! - DATA_HASH is the data hash of the non-fungible asset to build.
+#! - ASSET is the built non-fungible asset.
+export.build_non_fungible_asset
+    # assert the faucet is a non-fungible faucet
+    dup exec.internal_account::is_non_fungible_faucet assert
+    # => [faucet_id, DATA_HASH]
+
+    # build the asset
+    movup.3 drop movdn.2
+    # => [hash_0, hash_1, faucet_id, hash_3]
+
+    # set the most significant bit of the asset to 0
+    mul.2 div.2
+    # => [ASSET]
```

Comment:
I didn't catch this in the previous review, but I don't think this will work. The issue is that we are doing field multiplications here, so `mul.2` should just wrap around and then `div.2` would bring us back to exact same value.

To clear the most significant bit we'd need to split the element into 32-bit limbs, then use a u32 division to clear the bit, and then finally we'd need to re-combine the limbs.

Reply:
good catch!  This was my initial implementation but I changed it for a simpler implementation without consideration the implication of the field arithmetic.  I suppose the tests are passing as the asset we are using in the test has a 0 bit for the most significant bit.